core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java

```diff
@@ -243,11 +272,240 @@ protected AnalyzeResponse shardOperation(AnalyzeRequest request, ShardId shardId
                 IOUtils.closeWhileHandlingException(stream);
             }
         }
+        return tokens;
+    }
 
-        if (closeAnalyzer) {
-            analyzer.close();
+    private DetailAnalyzeResponse detailAnalyze(AnalyzeRequest request, Analyzer analyzer, String field) {
+        DetailAnalyzeResponse detailResponse;
+        final Set<String> includeAttributes = Sets.newHashSet();
+        if (request.attributes() != null && request.attributes().length > 0) {
+            for (String attribute : request.attributes()) {
+                includeAttributes.add(attribute.toLowerCase(Locale.ROOT));
+            }
+        }
+
+        CustomAnalyzer customAnalyzer = null;
+        if (analyzer instanceof CustomAnalyzer) {
+            customAnalyzer = (CustomAnalyzer)analyzer;
+        } else if (analyzer instanceof NamedAnalyzer && ((NamedAnalyzer) analyzer).analyzer() instanceof CustomAnalyzer) {
+            customAnalyzer = (CustomAnalyzer) ((NamedAnalyzer) analyzer).analyzer();
+        }
+
+        if (customAnalyzer != null) {
+            // customAnalyzer = divide charfilter, tokenizer tokenfilters
+            CharFilterFactory[] charfilterFactories = customAnalyzer.charFilters();
+            TokenizerFactory tokenizerFactory = customAnalyzer.tokenizerFactory();
+            TokenFilterFactory[] tokenfilterFactories = customAnalyzer.tokenFilters();
+
+            Map<String, List<String>> charFiltersTexts = Maps.newHashMapWithExpectedSize(charfilterFactories.length);
+            Map<String, TokenListCreator> tokenFiltersTokenListCreator = Maps.newHashMapWithExpectedSize(tokenfilterFactories.length);
+
+            TokenListCreator tokenizerTokenListCreator = new TokenListCreator();
+
+            for (String text : request.text()) {
+                String charFilteredSource = text;
+
+                Reader reader = new FastStringReader(text);
+                if (charfilterFactories != null) {
+                    for (CharFilterFactory charfilter : charfilterFactories) {
+                        reader = charfilter.create(reader);
+                        Reader readerForWriteOut = new FastStringReader(charFilteredSource);
+                        readerForWriteOut = charfilter.create(readerForWriteOut);
+                        charFilteredSource = writeCharStream(readerForWriteOut);
+
+                        List<String> texts = charFiltersTexts.get(charfilter.name());
+                        if (texts == null) {
+                            texts = Lists.newArrayList();
+                        }
+                        texts.add(charFilteredSource);
+                        charFiltersTexts.put(charfilter.name(), texts);
```

Comment:
Wondering if you want to put again the `texts` object in the map.
I would probably do something like:

```java
List<String> texts = charFiltersTexts.get(charfilter.name());
if (texts == null) {
    texts = Lists.newArrayList();
    charFiltersTexts.put(charfilter.name(), texts);
}
texts.add(charFilteredSource);
```